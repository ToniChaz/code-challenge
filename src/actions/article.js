import type from '../constants/actions/article'

export default {
  fetchArticles: () => ({ type: type.FETCH_ARTICLES }),
  fetchArticlesSuccess: articles => ({ type: type.FETCH_ARTICLES_SUCCESS, articles }),
  fetchArticlesError: error => ({ type: type.FETCH_ARTICLES_ERROR, error })
}
