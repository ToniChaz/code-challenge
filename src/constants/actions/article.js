export default {
  FETCH_ARTICLES: 'FETCH_ARTICLES',
  FILTER_ARTICLE: 'FILTER_ARTICLE',
  DELETE_ARTICLE: 'DELETE_ARTICLE',
  UPDATE_ARTICLE: 'UPDATE_ARTICLE',
  CREATE_ARTICLE: 'CREATE_ARTICLE',
  DELETE_ARTICLE_SUCCESS: 'DELETE_ARTICLE_SUCCESS',
  FETCH_ARTICLES_SUCCESS: 'FETCH_ARTICLES_SUCCESS',
  DELETE_ARTICLES_SUCCESS: 'DELETE_ARTICLES_SUCCESS',
  UPDATE_ARTICLE_SUCCESS: 'UPDATE_ARTICLE_SUCCESS',
  CREATE_ARTICLE_SUCCESS: 'CREATE_ARTICLE_SUCCESS'
}
