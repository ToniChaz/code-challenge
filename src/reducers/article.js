import type from '../constants/actions/article'

export function initialState() {
  return {
    articles: null,
    loading: true,
    error: null
  }
}

function fetchArticles(state) {
  return {
    ...state,
    loading: true,
    error: null
  }
}

function fetchArticlesSuccess(state, { articles }) {
  return {
    ...state,
    articles,
    loading: false,
    error: null
  }
}

function fetchArticlesError(state, { error }) {
  return {
    ...state,
    loading: false,
    error
  }
}

export default function(state = initialState(), action) {
  switch (action.type) {
    case type.FETCH_ARTICLES:
      return fetchArticles(state, action)
    case type.FETCH_ARTICLES_SUCCESS:
      return fetchArticlesSuccess(state, action)
    case type.FETCH_ARTICLES_ERROR:
      return fetchArticlesError(state, action)
    default:
      return state
  }
}
