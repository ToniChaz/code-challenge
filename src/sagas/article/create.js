import { call, put, takeEvery } from 'redux-saga/effects'
import { push } from 'react-router-redux'
import articleService from '../../services/article/create'
import action from '../../actions/article'
import alertAction from '../../actions/alert'
import modalAction from '../../actions/modal'
import type from '../../constants/actions/article'
import { CREATE } from '../../constants/service/mutations'

export function* createArticle({ article }) {
  try {
    const response = yield call(articleService, CREATE, article)
    yield put(action.createArticleSuccess(response.data.createArticle))
    yield put(alertAction.openAlert('success', 'The article has been saved successfully'))
    yield put(modalAction.closeModal())
    yield put(push('/'))
  } catch (error) {
    yield put(alertAction.openAlert('danger', `An error occurred: ${error.message}`))
  }
}

export function* watchCreateArticle() {
  yield takeEvery(type.CREATE_ARTICLE, createArticle)
}
