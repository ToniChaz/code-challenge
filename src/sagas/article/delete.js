import { call, put, takeEvery } from 'redux-saga/effects'
import { push } from 'react-router-redux'
import articleService from '../../services/article/delete'
import action from '../../actions/article'
import alertAction from '../../actions/alert'
import type from '../../constants/actions/article'
import { DELETE } from '../../constants/service/mutations'

export function* deleteArticle({ articleId }) {
  try {
    const response = yield call(articleService, DELETE, articleId)
    yield put(action.deleteArticleSuccess(response.data.deleteArticle.id))
    yield put(alertAction.openAlert('success', 'The article has been deleted successfully'))
    yield put(push('/'))
  } catch (error) {
    yield put(alertAction.openAlert('danger', `An error occurred: ${error.message}`))
  }
}

export function* watchDeleteArticle() {
  yield takeEvery(type.DELETE_ARTICLE, deleteArticle)
}
