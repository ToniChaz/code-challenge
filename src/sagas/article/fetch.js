import { call, put, takeEvery } from 'redux-saga/effects'
import articleService from '../../services/article/fetch'
import action from '../../actions/article'
import alertAction from '../../actions/alert'
import type from '../../constants/actions/article'
import { ARTICLES_QUERY } from '../../constants/service/queries'

export function* fetchArticles() {
  try {
    const response = yield call(articleService, ARTICLES_QUERY)
    yield put(action.fetchArticlesSuccess(response.data.articles))
  } catch (error) {
    yield put(alertAction.openAlert('danger', `An error occurred: ${error.message}`))
  }
}

export function* watchFetchArticles() {
  yield takeEvery(type.FETCH_ARTICLES, fetchArticles)
}
