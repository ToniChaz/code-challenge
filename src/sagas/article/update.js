import { call, put, takeEvery } from 'redux-saga/effects'
import articleService from '../../services/article/update'
import action from '../../actions/article'
import alertAction from '../../actions/alert'
import modalAction from '../../actions/modal'
import type from '../../constants/actions/article'
import { UPDATE } from '../../constants/service/mutations'

export function* updateArticle({ article }) {
  try {
    const response = yield call(articleService, UPDATE, article)
    yield put(action.updateArticleSuccess([response.data.updateArticle]))
    yield put(alertAction.openAlert('success', 'The article has been updated successfully'))
    yield put(modalAction.closeModal())
  } catch (error) {
    yield put(alertAction.openAlert('danger', `An error occurred: ${error.message}`))
  }
}

export function* watchUpdateArticle() {
  yield takeEvery(type.UPDATE_ARTICLE, updateArticle)
}
