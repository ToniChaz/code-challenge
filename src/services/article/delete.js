import { requestHandler } from '../serviceHandler'

export default function(mutation, articleId) {
  return requestHandler('/graphql', mutation, { id: articleId })
}
