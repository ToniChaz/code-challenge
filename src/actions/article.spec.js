// article.spec.js

import article from './article'
import { article as articleMock } from '../../test/mocks'

describe('Article action', () => {
  test('fetch', () => {
    expect(article.fetchArticles()).toMatchSnapshot()
  })

  test('filterArticle', () => {
    expect(article.filterArticle(
      { articleId: '59eb8261d2ddc040ead6c7fa' },
      () => 'resolve',
      () => 'reject'
    )).toMatchSnapshot()
  })

  test('deleteArticle', () => {
    expect(article.deleteArticle(
      { articleId: '59eb8261d2ddc040ead6c7fa' },
      () => 'resolve',
      () => 'reject'
    )).toMatchSnapshot()
  })

  test('updateArticle', () => {
    expect(article.deleteArticle(
      { articleMock },
      () => 'resolve',
      () => 'reject'
    )).toMatchSnapshot()
  })

  test('createArticle', () => {
    expect(article.deleteArticle(
      { articleMock },
      () => 'resolve',
      () => 'reject'
    )).toMatchSnapshot()
  })
})
