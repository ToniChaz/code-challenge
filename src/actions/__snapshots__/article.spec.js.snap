exports[`Article action createArticle 1`] = `
Object {
  "articleId": Object {
    "articleMock": Array [
      Object {
        "_id": "59eb8261d2ddc040ead6c7fa",
        "author": "Bertram Wintheiser",
        "content": "Et iure enim saepe. Repellat reprehenderit sequi id. Eveniet sit ex. Rem et labore temporibus dicta architecto reprehenderit qui. Consequatur amet aspernatur porro. Quo consequatur nesciunt et quod id.",
        "excerpt": "Et iure enim saepe. Repellat reprehenderit sequi id. Eveniet sit ex. Rem et labore temporibus dicta ",
        "published": false,
        "tags": Array [
          "Avon portal Outdoors",
          "networks Berkshire Fish",
        ],
        "title": "Senior Communications Coordinator",
      },
    ],
  },
  "type": "DELETE_ARTICLE",
}
`;

exports[`Article action deleteArticle 1`] = `
Object {
  "articleId": Object {
    "articleId": "59eb8261d2ddc040ead6c7fa",
  },
  "type": "DELETE_ARTICLE",
}
`;

exports[`Article action fetch 1`] = `
Object {
  "type": "FETCH_ARTICLES",
}
`;

exports[`Article action filterArticle 1`] = `
Object {
  "articleId": Object {
    "articleId": "59eb8261d2ddc040ead6c7fa",
  },
  "type": "FILTER_ARTICLE",
}
`;

exports[`Article action updateArticle 1`] = `
Object {
  "articleId": Object {
    "articleMock": Array [
      Object {
        "_id": "59eb8261d2ddc040ead6c7fa",
        "author": "Bertram Wintheiser",
        "content": "Et iure enim saepe. Repellat reprehenderit sequi id. Eveniet sit ex. Rem et labore temporibus dicta architecto reprehenderit qui. Consequatur amet aspernatur porro. Quo consequatur nesciunt et quod id.",
        "excerpt": "Et iure enim saepe. Repellat reprehenderit sequi id. Eveniet sit ex. Rem et labore temporibus dicta ",
        "published": false,
        "tags": Array [
          "Avon portal Outdoors",
          "networks Berkshire Fish",
        ],
        "title": "Senior Communications Coordinator",
      },
    ],
  },
  "type": "DELETE_ARTICLE",
}
`;
